pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/bloom/BloomIndexType.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.Constants;
+import org.apache.pinot.segment.spi.V1Constants;
+import org.apache.pinot.segment.spi.creator.IndexCreationContext;
+import org.apache.pinot.segment.spi.index.FieldIndexConfigs;
+import org.apache.pinot.segment.spi.index.IndexHandler;
+import org.apache.pinot.segment.spi.index.IndexReaderFactory;
+import org.apache.pinot.segment.spi.index.IndexType;
+import org.apache.pinot.segment.spi.index.creator.BloomFilterCreator;
+import org.apache.pinot.segment.spi.index.reader.BloomFilterReader;
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.BloomFilterConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+
+public class BloomIndexType implements IndexType<BloomFilterConfig, BloomFilterReader, BloomFilterCreator> {
+  public static final BloomIndexType INSTANCE = new BloomIndexType();
+
+  @Override
+  public String getId() {
+    return "bloom_filter";
+  }
+
+  @Override
+  public Class<BloomFilterConfig> getIndexConfigClass() {
+    return BloomFilterConfig.class;
+  }
+
+  @Override
+  public BloomFilterConfig getDefaultConfig() {
+    return BloomFilterConfig.DISABLED;
+  }
+
+  @Override
+  public BloomFilterConfig getConfig(TableConfig tableConfig, Schema schema) {
+    throw new UnsupportedOperationException("To be implemented in a future PR");
+  }
+
+  @Override
+  public BloomFilterCreator createIndexCreator(IndexCreationContext context, BloomFilterConfig indexConfig) {
+    int cardinality = context.getCardinality();
+    if (cardinality == Constants.UNKNOWN_CARDINALITY) {
+      // This is when we're creating bloom filters for non dictionary encoded cols where exact cardinality is not
+      // known beforehand.
+      // Since this field is only used for the estimate cardinality, using total # of entries instead
+      // TODO (saurabh) Check if we can do a better estimate
+      cardinality = context.getTotalNumberOfEntries();
+    }
+    return new OnHeapGuavaBloomFilterCreator(context.getIndexDir(), context.getFieldSpec().getName(), cardinality,
+        indexConfig, context.getFieldSpec().getDataType());
+  }
+
+  @Override
+  public IndexReaderFactory<BloomFilterReader> getReaderFactory() {
+    return new IndexReaderFactory.Default<BloomFilterConfig, BloomFilterReader>() {
+      @Override
+      protected IndexType<BloomFilterConfig, BloomFilterReader, ?> getIndexType() {
+        return BloomIndexType.INSTANCE;
+      }
+
+      @Override
+      protected BloomFilterReader createIndexReader(PinotDataBuffer dataBuffer, ColumnMetadata metadata,
+          BloomFilterConfig indexConfig) {
+        return BloomFilterReaderFactory.getBloomFilterReader(dataBuffer, indexConfig.isLoadOnHeap());
```

Comment:
Can BloomFilterReaderFactory itself be an IndexReaderFactory implementation?

Reply:
Yes, we can do it. I just wanted to minimize the number of changes in order to make the PR easier to read, but we can modify that in `index-spi` branch. Please note that this PR includes `BloomIndexType` as just an example and the implementation doesn't actually matter that much, the next PR (https://github.com/apache/pinot/pull/10193) is where other indexes are created and later in https://github.com/apache/pinot/pull/10184 is where most of the changes actually may (or may not :D) make sense.